### PR TITLE
`%lick` vane

### PR DESCRIPTION
**Pull request by mopfel-winrux**
This is a PR to include a new vane called `%lick`. This vane allows `%gall` agents to open an IPC communication port on urth. This port speaks in jammed nouns of the form `[=mark =noun]`

The vane has three parts:

1. urbit/urbit modifications: This PR
2. urbit/vere modifications: https://github.com/urbit/vere/pull/358
3. userspace example: https://github.com/mopfel-winrux/lick-agent

TODO:

- [x] Change port name from `@tas` to `$path`
- [x] Write python code that connects to the agent-port and performs `jam`s and `cue`s on data
- [x] Write python code that connects to a HW device and acts as a middleman between the port and device
- [x] Gall should intercept all cards and prefix the agent name to the socket name. It should also remove the name for any `%soak`s

Add state to Gall to track which agent owns which socket (POSTPONED for gall rewrite)

Notes:

Useful links:
- https://github.com/urbit/urbit/blob/develop/pkg/arvo/sys/vane/gall.hoon#L1133 for modifying card
- https://github.com/urbit/urbit/blob/develop/pkg/arvo/sys/vane/gall.hoon#L602 for modifying soak
- https://github.com/urbit/urbit/blob/develop/pkg/arvo/sys/vane/gall.hoon#L1404 for modifying soak

Prefix agent to note and something to wire to note that its a lick command

Open Design Questions:
- Right now the owner of the ipc port is the last agent that `%spin`ed it. This should probably change to the first one.
- Any agent can `%spit` on an ipc port, even if they didn't %spin it. Do we want to gate `%spit`ing and `%shut`ing to only the agent that spun it?

### CODE REVIEW

**Comment by tacryt-socryp** [comment]
absolutely huge

File: pkg/arvo/sys/vane/lick.hoon [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      ::        $>(%deal task:gall)                       ::  full transmission
+      ::    ==                                            ::
+      ::==                                            ::
+      ::
+    +$  sign  ~
+    ::
+    +$  lick-state
+      $:  %0
+          unix-duct=duct
+          agents=(map name duct)
```
Comment (mopfel-winrux) [inline pkg/arvo/sys/vane/lick.hoon:20]: Change agent to better name:
listener
owner

File: pkg/arvo/sys/vane/lick.hoon [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          dud=(unit goof)
+          wrapped-task=(hobo task)
+      ==
+  ^-  [(list move) _lick-gate]
+  ::
+  =/  =task  ((harden task) wrapped-task)
+  ?+   -.task  [~ lick-gate]
+      %born     :: need to register devices with vere and send disconnect soak
+    :-  %+  weld 
+        (turn ~(tap in ~(key by agents.state)) register)
```
Comment (mopfel-winrux) [inline pkg/arvo/sys/vane/lick.hoon:60]: indent

File: pkg/arvo/sys/vane/lick.hoon [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        (turn ~(tap in ~(key by agents.state)) register) 
+        (turn ~(tap in ~(key by agents.state)) disconnect)
+    lick-gate(unix-duct hen) 
+    ::
+      %spin     :: A gall agent wants to spin a communication line
+    :-  ~[(register name.task)]
+    lick-gate(agents (~(put by agents) name.task hen))
+    ::
+      %shut     :: shut down a communication line
+    :-  ~[[unix-duct.state %give [%shut name.task]]]
```
Comment (mopfel-winrux) [inline pkg/arvo/sys/vane/lick.hoon:69]: trailing sig instead of double brackets

File: pkg/arvo/sys/vane/lick.hoon [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+++  load
+  |=  old=lick-state
+  ^+  lick-gate
+  ~&  >>  "lick load:"
+  lick-gate(state old)
+::  +scry: view state
+::
+::  %a  scry out a list of devices
+::  %d  get a device's  dat and tus
+++  scry
```
Comment (mopfel-winrux) [inline pkg/arvo/sys/vane/lick.hoon:92]: Implement scry for a list of ipc ports and a ports owner

File: pkg/arvo/sys/vane/lick.hoon [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  =*  lot=coin  $/r.bem
+  =*  tyl  s.bem
+  ~&  >>  "lick scry:" 
+  ~&  >>  ["lyc" lyc]
+  ~&  >>  ["car" car]
+  ~&  >>  ["bem" bem]
+  ~
+::
+++  stay  
+  ~&  >>  "lick stay:"
```
Comment (mopfel-winrux) [inline pkg/arvo/sys/vane/lick.hoon:108]: Cleanup sigpams

File: pkg/arvo/sys/vane/lick.hoon [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  ?+   -.task  [~ lick-gate]
+      %born     :: need to register devices with vere and send disconnect soak
+    :-  %+  weld 
+          (turn ~(tap in ~(key by owners.state)) register) 
+        (turn ~(tap in ~(key by owners.state)) disconnect)
+    lick-gate(unix-duct hen) 
+    ::
+      %spin     :: A gall agent wants to spin a communication line
+    ::  XX Right now the last agent to register a port owns it. Should
+    ::  this change?
```
Comment (matthew-levan) [inline pkg/arvo/sys/vane/lick.hoon:61]: Can this just be first come first served? As long as the agent is installed (and not suspended?) it retains ownership, but once uninstalled relinquishes it.

File: pkg/arvo/sys/vane/ames.hoon [hunk 7]
```diff
@@ -4449,7 +4453,7 @@
             ::  if this was a re-send, don't adjust rtt or downstream state
             ::
             ?:  (gth tries.packet-state 1)
-              metrics
+              metrics(rto (clamp-rto (add rtt (mul 4 rttvar))))
```
Comment (matthew-levan) [inline pkg/arvo/sys/vane/ames.hoon:4456]: What is this doing?

Comment (mopfel-winrux) [reply pkg/arvo/sys/vane/ames.hoon:4456]: This is a merge issue I between develop and 413 branch I did not touch this

File: pkg/arvo/sys/vane/clay.hoon
Comment (matthew-levan) [inline pkg/arvo/sys/vane/clay.hoon:1]: Are these changes part of your PR or just another merge error? I'm working on some of these minor fixes for you in a separate branch which I'll open a PR for, hope it helps.

File: pkg/arvo/sys/vane/gall.hoon
Comment (matthew-levan) [inline pkg/arvo/sys/vane/gall.hoon:1]: Same question here.

File: pkg/arvo/sys/vane/clay.hoon
Comment (mopfel-winrux) [reply pkg/arvo/sys/vane/clay.hoon:1]: Another merge error. I had my branch based on develop and 412 isnt based on develop

File: pkg/arvo/sys/vane/ames.hoon
Comment (mopfel-winrux) [inline pkg/arvo/sys/vane/ames.hoon:1]: I did not modify this file

File: pkg/arvo/sys/vane/clay.hoon
Comment (mopfel-winrux) [inline pkg/arvo/sys/vane/clay.hoon:1]: I did not modify this file

**Comment by mopfel-winrux** [comment]
There seems to be an issue with the %born card being sent twice. I think it has to do with the gall stuff I added. This was discovered during the Build party and I didn't have time to debug it, Making a note of it here so I can debug it next week.

**Comment by mopfel-winrux** [comment]
> There seems to be an issue with the %born card being sent twice. I think it has to do with the gall stuff I added. This was discovered during the Build party and I didn't have time to debug it, Making a note of it here so I can debug it next week.

This was just a state bug due to the modifications made to gall. It does not show up on a new install.

**Review by joemfb — APPROVED** [review]
this will do

**Review by joemfb — CHANGES_REQUESTED** [review]
CI is fixed now, but it's still failing to build this vane.

**Comment by joemfb** [comment]
CI is failing due to #6605 in the target branch. %lick's `+scry` arm needs to be updated to the new interface.

**Review by joemfb — APPROVED** [review]

**Pull request merged**